### DISCUSSION TITLE
Day 2 is out!

### DISCUSSION THREAD

**Discussion by CarolineMorton** [CarolineMorton/AdventOfCode_WIR#3] [Ideas]
🎅 **Day 2 is out!** 🎄

Following on from our lovely kick off yesterday, day 2 is here. We are letting people have 24 hours before posting a solution directly into the forum. Please feel free to post links out to your repos before then - then it is up to readers if they choose to click on them or not. 

I've already seen some create solutions to Day 1 - particularly got to highlight [this](https://github.com/hollygrimm/AdventOfCode_WIR/blob/b71aa35ff6b4cedfcea64b4ab1e1b5c5d6d251ee/day_01/src/main.rs#L34) from @hollygrimm who spend some time implementing some QC checks and custom errors using enums. This is my preferred solution for error checking these days. 

Day 2 presents a perfect opportunity to explore:

- Processing variable-length data structures
- Working with sequences and patterns

See you all in the comments

Caroline

**Comment by hollygrimm** [comment]
My [day 2 solution](https://github.com/hollygrimm/AdventOfCode_WIR/blob/main/day_02/src/main.rs)

**Comment by em-baggie** [comment]
My <a href = https://github.com/em-baggie/AdventOfCode_WIR/blob/main/day_02/src/main.rs>solution</a> for day 2 🦀

**Reply by hollygrimm** [reply]
I love your use of the windows method in your solution

**Reply by em-baggie** [reply]
> I love your use of the windows method in your solution
 
Thanks! I initially had a solution with a load of for loops but it was super long and clunky. I wondered if it was possible to convert the vec into slices and analyse the them separately without allocating new memory. I came across the windows method after a bit of googling!

**Comment by CarolineMorton** [comment]
My solution is here: https://github.com/CarolineMorton/AdventOfCode_WIR/tree/caroline-solutions/day_02

I made extensive use of test cases to check that the logic was functioning as it should be. I used the example data in the brief for most of these tests - extended a little. I also used command line args to pass the file path in. 

It could be improved with QC of the input file before any iterating and custom errors.

**Reply by CarolineMorton** [reply]
This really shows my preference for `main.rs` files. I prefer my main.rs files to be quite short and snappy so I defined the functionality i need, with unit tests in other files, and imported them in. This left my main fn quite clean: 

```rust
fn main() {
    println!("Welcome to Day 2!");

    // Parse command line arguments
    let args = Args::parse();

    // Load the reports from the file
    let reports = load_reports(&args.file);

    // Check the reports without problem dampener on
    let (safe_reports, _) = check_reports(&reports, false);

    // Print the safe reports length with problem dampener off
    println!("Number of safe reports without problem dampener: {}", safe_reports.len());

    // Check the reports with problem dampener on
    let (safe_reports, _) = check_reports(&reports, true);

    // Print the safe reports length with problem dampener on
    println!("Number of safe reports with problem dampener: {}", safe_reports.len());
}
```

Be interested to see what other people think of this approach!

**Reply by hollygrimm** [reply]
I think the separate files and modules make sense. It's much more organized, especially for testing. And easy to reuse them for the next solution.

**Reply by em-baggie** [reply]
Agree, main file shows the flow nicely and abstracts away the detailed implementation

**Comment by Sowmiya07** [comment]
[Here](https://github.com/Sowmiya07/AdventOfCode_WIR/blob/main/day_02/src/main.rs) is my Day 2 solution

**Reply by CarolineMorton** [reply]
Well done for getting this done! 

Would you like a review of your code?

**Reply by Sowmiya07** [reply]
Thanks @CarolineMorton. Yes, I would love that. I am a beginner and learning from review would be really helpful.

**Reply by CarolineMorton** [reply]
### Great Work on the Code! 🌟

I particularly like that you have well-labeled functions that explain what they do. I am a big fan of having well described functions. This makes the code easier to follow. Breaking the logic into smaller, modular functions is excellent—it not only makes the code reusable but also sets you up nicely for testing (see below). 

### Things to Consider for Improvement ✨

#### 1. **Adding Tests**
Including tests at the bottom of your code allows you to verify that each function behaves as expected. Rust has great support for testing using `cargo test`, and has the format thaat you write your unit tests in the file that you are testing typically rather than other languages (like python) where it might be  in its own folder called tests. So for example:

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_remove_ith_level() {
        let levels = vec![10, 20, 30, 40];
        assert_eq!(remove_ith_level(&levels, 1), vec![10, 30, 40]);
    }
}
```

#### 2. **Avoid Panic for Control Flow**
While `panic!` is fine for debugging or catching serious issues, using it for normal control flow isn’t ideal. Consider handling cases like missing levels more gracefully by returning a `Result` or an `Option`. This avoids crashing the program if the input data isn’t perfect and provides more meaningful error messages.

#### 3. **Adding Docstrings***
Adding brief docstrings to your functions will improve readability. For example:

```rust
/// Checks whether the report is increasing, decreasing, or invalid.
/// 
/// # Arguments
/// * `levels` - A slice of integers representing safety levels.
///
/// # Returns
/// A `u16` where:
/// - 1 indicates a valid report (increasing or decreasing)
/// - 0 indicates an invalid report
fn check_report(levels: &[u16]) -> u16 {
    // Function body
    ....
}
```

**Reply by Sowmiya07** [reply]
Thank you @CarolineMorton for taking time to review the code and give us detailed comments on how to improve. I will consider all these pointers for upcoming problems 💯

**Reply by CarolineMorton** [reply]
Great work here! :-)

**Comment by DominikStefancik** [comment]
[My solution](https://github.com/DominikStefancik/AdventOfCode_WIR/tree/day_02/day_02) for the day 2

**Reply by CarolineMorton** [reply]
Nice one - how did you find it? 

I noticed that you are using `lib.rs` as the main container for your code - I'm curious if that was a deliberate choice? I've worked with multi-workspace projects before where lib.rs played a key role in code sharing / compiling.

**Reply by DominikStefancik** [reply]
The first part was relatively easy. But for some reason the second part took me longer than I expected. I tried different approaches, but they didn't work to my surprise 😄 
Yeah, the `lib.rs` was a deliberate choice. The code was growing and I didn't want to have everything in main.rs. I didn't think about the aspect of sharing the code when putting it there. Maybe you are right, I could have named the file as `report_stats.rs` instead. Will keep that in mind next time. Thanks for bringing it up 🙏

**Reply by CarolineMorton** [reply]
No worries. I work on a large code base with lots of `lib.rs` files dotted around in different workspaces so it jumped out to me. Happy to share that at some point if helpful.  
Well done on this!

**Comment by kuhmil** [comment]
[Day 2](https://github.com/kuhmil/AdventOfCode_WIR/tree/main/day_02)!

**Reply by CarolineMorton** [reply]
Awesome work here @kuhmil  - great to see use of `Result()` in particular

**Comment by ladame** [comment]
Oh.....I am so late....... Anyway, I try to catch you up !!!
This is my [solution](https://github.com/ladame/AdventOfCode_WIR/tree/main/day_02).
I try to apply TDD and clean code principles.
